### PR TITLE
organize project files

### PR DESCRIPTION
**Pull request by elmeraa-usgs**
I dded src and out folders within each phase folder; moved my_happy_script.R to src folder of 1_fectch following issue #3 steps.

### CODE REVIEW

**Comment by elmeraa-usgs** [comment]
Followed along with issue #4 in creating functions to generate necessary output data and visualization. Let me know if there's a better way to streamline this, thanks!

**Comment by padilla410** [comment]
## Overview
Howdy Elmera,

I would like to do this review in two big rounds of edits. In the first round I'm going to focus on structure and setting up this repo to make it easy for your reviewer (big stuff) and the second round will be the nitty gritty of the functions themselves (little stuff).

So, here are some of the overarching big stuff comments and how I would like you to address them:

## Big stuff first

### I cannot run your functions as written

The main reason for this is because you have added `*/out/*` to the `.gitignore` file. This was the correct thing to do, but I would like to see something in your code that checks for the existence of the `out` subfolder. In the original `my_happy_script.R` that happens [here](https://github.com/elmeraa/ds-pipelines-targets-1/blob/3b2a18eef39d000306adbee5252520266b05c444/my_work_R/my_happy_script.R#L10). Adding this functionality into your script will save the reviewer a little troubleshooting on their end. You can either modify your existing functions to check for the appropriate directories or you can write a stand-alone function that does this.

To help out your reviewer, I would also recommend that you add a file at the top level of the directory called `00_run.R` or something like that. Having a file like that would let the reviewer quickly run all your functions to verify that they work instead of having to open each individual `<<phase>>/src/<<file_name.R>>`. It would also let you consolidate all of your `library` calls to one central location, instead of adding them to each function (see [here](https://github.com/elmeraa/ds-pipelines-targets-1/blob/1bc78cc367dcc288cf540ba38ca8131f94ba0331/1_fetch/src/fetch.R#L5-L9), [here](https://github.com/elmeraa/ds-pipelines-targets-1/blob/1bc78cc367dcc288cf540ba38ca8131f94ba0331/2_process/src/process.R#L5-L9), and [here](https://github.com/elmeraa/ds-pipelines-targets-1/blob/1bc78cc367dcc288cf540ba38ca8131f94ba0331/3_visualize/src/visualize.R#L5-L10))   

### Add documentation to your functions

There is this saying that "your closest collaborator is yourself from 6 months ago". So to help yourself and others out all functions should have documentation. When working in R I use a modified version of the [Roxygen](https://style.tidyverse.org/documentation.html) format. You should always include the following components:
* A description of what the function does (some times this can be as simple as a title, other times you need a bit more)
* A description of every function argument, including information about the expected `class`. This can help someone else debug quickly

Here are some examples from my work:
* [`subsset_model_obs_data`](https://github.com/USGS-R/lake-temperature-missouri-models/blob/d45326368103b9eb6993eab15cc381826760847b/1_prep/src/munge_obs.R#L1-L6)
* [`create_pest_raster`](https://code.usgs.gov/wma/iidd/pnsp-pipeline/-/blob/main/2_process/src/create_pest_raster.R#L1-8)

### Consider adding more detail to yor PR writeups

I don't need a commit related to this topic, but I do want you to think about it for the rest of your PRs. PRs do two things: (1) they should provide enough context for the reviewer to understand what you are doing and why, and (2) they serve as documentation for your decision-making process. They can take many forms, but I usually try to include the following headings:
1. Overview - **why** are you opening this PR? Is it related to an issue? Some other reason? Please make this clear.
1. Summary of Changes - this is a broad category where I usually talk about the overarching reasons for big changes
1. Verification of results - This might not be relevant for this training, but I like to include a section where I prove to my reviewer that my code does what I said it would. I have caught several mistakes this way.

I'm going to link you to a few example PRs to describe what I mean:
* [PNSP pipeline](https://code.usgs.gov/wma/iidd/pnsp-pipeline/-/merge_requests/14)
* [Lake temperature modeling parsers, PR 337](https://github.com/USGS-R/lake-temperature-model-prep/pull/337)
* [lake temp parsers](https://github.com/USGS-R/lake-temperature-model-prep/pull/321) this one is short and sweet because the "why" was already documented in several issues

### Missing content?

Using your updated functions, were you able create all 4 of the expected files?:
* figure_1.png
* model-diagnostic_text.R
* model_RMSEs.csv
* model_summary_results.csv

As I mentioned, i wasn't able to run the code yet, but I don't see [this](https://github.com/elmeraa/ds-pipelines-targets-1/blob/3b2a18eef39d000306adbee5252520266b05c444/my_work_R/my_happy_script.R#L80-L96) chunk in any of your functions. You may need to add another function.



I would like to see some commits that address the comments above, and then we will work our way through the functions themselves.

**Comment by elmeraa-usgs** [comment]
Thanks for the breakdown and providing examples, Julie! 

### Overview
The purpose of this PR is to integrate Julie's notes related to issue #4 of the targets pipeline training. 

### Summary of changes:
- added ` 00_run.R ` which includes necessary libraries, and sources all four functions
- added `diagnostics.R` which exports a `model_diagnostic_text.txt` file that includes the mean RMSEs for the three model types as well as a comparative breakdown of training temperature profiles for the model types.  
- edited `fetch.R` , `process.R`, `visualize.R`, and `diagnostics.R` to include the Roxygen skeleton with more details on what each function exports. For each of the listed files I also added a `dir.exists()` to check for the existence of the `out` subfolder, as you suggested.

File: 1_fetch/src/fetch.R [hunk 1]
```diff
@@ -0,0 +1,17 @@
+#' Fetch data function
+#'
+#' @export model_RMSE csv, file includes exper_n, exper_id, model_type & rmse
+
+fetch <- function(){
+  
+  # Check for existence of out subfolder
+  dir.exists("1_fetch/out")
```
Comment (padilla410) [inline 1_fetch/src/fetch.R]: Good! Now it needs to do something if it doesn't find the directory. Right now running this line returns `FALSE` for me.

File: 1_fetch/src/fetch.R [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  
+  # Get the data from ScienceBase
+  item_file_download('5d925066e4b0c4f70d0d0599', names = 'me_RMSE.csv', 
+                     destinations = '1_fetch/out/model_RMSEs.csv', overwrite_file = TRUE)
+  
+}
+
+# run fetch function
+fetch()
```
Comment (padilla410) [inline 1_fetch/src/fetch.R]: Move this to the body of `00_run.R`

File: 2_process/src/process.R [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  
+  # Check for existence of out subfolder
+  dir.exists("2_process/out")
+  
+  eval_data <- readr::read_csv('1_fetch/out/model_RMSEs.csv', col_types = 'iccd') %>%
+    filter(str_detect(exper_id, 'similar_[0-9]+')) %>%
+    mutate(col = case_when(
+      model_type == 'pb' ~ '#1b9e77',
+      model_type == 'dl' ~'#d95f02',
+      model_type == 'pgdl' ~ '#7570b3'
```
Comment (padilla410) [inline 2_process/src/process.R]: Theses colors are good candidates for conversion to an argument

File: 2_process/src/process.R [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    ), pch = case_when(
+      model_type == 'pb' ~ 21,
+      model_type == 'dl' ~ 22,
+      model_type == 'pgdl' ~ 23
```
Comment (padilla410) [inline 2_process/src/process.R]: These are also good candidates for an argument

**Comment by padilla410** [comment]
Nice job on the re-org, Elmera.

Now we are going to look inside your functions and do a little clean up. There are two areas that we are going to focus on as I add comments to your code:
1. What pieces of information are good candidates for conversion into a function argument?
2. Are there any pieces of our function bodies that are repetitious? If so, can we turn them into a function?

Also, I think ScienceBase might be down, so I checked `1_fetch/src/fetch.R` by comparing the code to my own. I then copied the data from my own local repo into this one.

File: 2_process/src/process.R [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    ), n_prof = as.numeric(str_extract(exper_id, '[0-9]+')))
+  
+  # export eval data
+  saveRDS(eval_data, file = '2_process/out/eval_data')
+  write_csv(eval_data, file = '2_process/out/model_summary_results.csv')
+}
+
+# run process function 
+process()
```
Comment (padilla410) [inline 2_process/src/process.R]: Move this to the body of `00_run.R`

File: 3_visualize/src/diagnostics.R [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+  ({{dl_500mean}} and {{pb_500mean}}°C, respectively) or more, but worse than PB when training was reduced to 100 profiles ({{dl_100mean}} and {{pb_100mean}}°C respectively) or fewer.
+  The PGDL prediction accuracy was more robust compared to PB when only two profiles were provided for training ({{pgdl_2mean}} and {{pb_2mean}}°C, respectively). '
+  
+  whisker.render(template_1 %>% str_remove_all('\n') %>% str_replace_all('  ', ' '), render_data ) %>% cat(file = file.path('3_visualize/out/model_diagnostic_text.txt'))
+  
+  
+}
+
+# run diagnostics function
+diagnostics()
```
Comment (padilla410) [inline 3_visualize/src/diagnostics.R]: Move this to the body of `00_run.R`

File: 3_visualize/src/visualize.R [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+  points(2.2, 1.09, col = '#1b9e77', pch = 21, bg = 'white', lwd = 2.5, cex = 1.5)
+  text(2.3, 1.1, 'Process-Based', pos = 4, cex = 1.1)
+  
+  dev.off()
+  
+  
+}
+
+# run visualization function
+visualize()
```
Comment (padilla410) [inline 3_visualize/src/visualize.R]: Move this to the body of `00_run.R`

File: 3_visualize/src/diagnostics.R [hunk 8]
```diff
@@ -0,0 +1,34 @@
+#' Export model diagnostics
+#' @export model_diagnostics_text txt, file contains mean RSME for three model types as well as a breakdown of training temperature profiles for the different models 
+
+diagnostics <- function(){
+  
+  # Check for existence of out subfolder
+  dir.exists("3_visualize/out")
+  
+  # Read in eval data 
+  eval_data <- readRDS("2_process/out/eval_data")
```
Comment (padilla410) [inline 3_visualize/src/diagnostics.R]: I'm calling it out here, but this comment applies to all the functions you've written - `in` and `out` files are great candidates for arguments. it's nice to have them as easy access so you can quickly change around file structure

File: 3_visualize/src/diagnostics.R [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+                      dl_980mean = filter(eval_data, model_type == 'dl', exper_id == "similar_980") %>% pull(rmse) %>% mean %>% round(2),
+                      pb_980mean = filter(eval_data, model_type == 'pb', exper_id == "similar_980") %>% pull(rmse) %>% mean %>% round(2),
+                      dl_500mean = filter(eval_data, model_type == 'dl', exper_id == "similar_500") %>% pull(rmse) %>% mean %>% round(2),
+                      pb_500mean = filter(eval_data, model_type == 'pb', exper_id == "similar_500") %>% pull(rmse) %>% mean %>% round(2),
+                      dl_100mean = filter(eval_data, model_type == 'dl', exper_id == "similar_100") %>% pull(rmse) %>% mean %>% round(2),
+                      pb_100mean = filter(eval_data, model_type == 'pb', exper_id == "similar_100") %>% pull(rmse) %>% mean %>% round(2),
+                      pgdl_2mean = filter(eval_data, model_type == 'pgdl', exper_id == "similar_2") %>% pull(rmse) %>% mean %>% round(2),
+                      pb_2mean = filter(eval_data, model_type == 'pb', exper_id == "similar_2") %>% pull(rmse) %>% mean %>% round(2))
```
Comment (padilla410) [inline 3_visualize/src/diagnostics.R]: This is a good chunk of repetitious code. I would like you to write a function to cut down on the copy-paste

File: 3_visualize/src/visualize.R [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+
+visualize <- function(){
+  
+  # Check for existence of out subfolder
+  dir.exists("3_visualize/out")
+  
+  # Read in eval data
+  eval_data <- readRDS("2_process/out/eval_data")
+  
+  png(file = file.path('3_visualize/out/figure_1.png'), width = 8, height = 10, res = 200, units = 'in')
```
Comment (padilla410) [inline 3_visualize/src/visualize.R]: Similar to colors and `pch`, these plot variables (width, height, res, units) are good candidates for arguments

**Review by padilla410 — CHANGES_REQUESTED** [review]

**Comment by elmeraa-usgs** [comment]
### Overview
Followed up on Julie's requested changes. The purpose of this PR is to continue with issue #4  of the targets pipeline training.

### Summary of changes:
- updated `dir.exists()` code within each function to returns `false` if directory already exists and `true` if it does not but was successfully created.
- moved the execution of the `fetch()` , `process()`, `visualize()` & `diagnostics()` functions to `00_run.R` .
- added `file_in = ` and `file_out = ` arguments for each function.
- added `col = ` and `pch = ` arguments to `process()` in `00_run.R` and edited `proccess.R` code to coordinate.
- added plot variables (`width`, `height`, `res`, & `units`) to `visualize()` arguement in `00_run.R`
- updated `diagnostics.R` to cut down on repetitious code. Here I used `sapply()` that creates a list of models and their associated means.

**Review by padilla410 — APPROVED** [review]
Nice job! I was able to run everything in `00_run.R` without a problem.

I have two general comments:
1. Double check your functions to make sure they're all documented
2. I would like to see a few more commits as you work through your PRs. When we program stuff we don't often work in a linear fashion, but the flexibility of staging files in `git` before committing let's us organize modified files in a logical manner. It also helps your reviewer track what you're up to (e.g., with well organized commits can help the reviewer verify that you've made all the requested changes). It's a little harder to track when it's one big ole commit (even though you did name it well).

You can merge this when you're ready.

**Pull request merged**